include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -523,6 +602,16 @@ class __future : private std::tuple<_Args...>
         return __buf.get_host_access(sycl::read_only)[0];
     }
 
+    template <typename _ExecutionPolicy, typename _T>
+    constexpr auto
+    __wait_and_get_value(__storage<_ExecutionPolicy, _T>& __buf)
+    {
+        // Explicit wait in case of USM memory. Buffer accessors are synchronous.
+        if (__buf.get_usm())
+            wait();
+        return __buf.get_value();
```

Comment:
SYCL buffer guarantees just data synchronization... But a kernel might still work...

Reply:
I think the data dependence should be on the buffer that is used in the kernel (read/write access). Thus, the host accessor called here should wait for the kernel completion. This follows the existing `__wait_and_get_value(sycl::buffer<_T>& __buf)` implementation above.

Reply:
"Thus, the host accessor called here should wait for the kernel completion"

I don't think so. The host accessor is waiting for the data will be copied back to the host, nothing more.

Reply:
Actually, looking at the spec:
```
As said in [Section 3.9.8](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:synchronization), the only blocking operations in SYCL (apart from explicit wait operations) are:

host accessor constructor, which waits for any kernels enqueued before its creation that write to the corresponding object to finish and be copied back to host memory before it starts processing. The host accessor does not necessarily copy back to the same host memory as initially given by the user;
```

So,